File: BddPipe/BddPipe/StepResultExtensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using BddPipe.Model;

namespace BddPipe
{
    internal static class StepResultExtensions
    {
        private const int DefaultIndentSize = 2;

        private static Some<string> ToOutcomeText(this Outcome outcome)
        {
            switch (outcome)
            {
                case Outcome.Pass: return "Passed";
                case Outcome.Fail: return "Failed";
                case Outcome.Inconclusive: return "Inconclusive";
                case Outcome.NotRun:
                default: return "not run";
            }
        }

        private static int ToIndentSizeStepAndBut(this Step step) =>
            step == Step.And || step == Step.But
                ? DefaultIndentSize
                : 0;

        private static int ToIndentSizeStep(this bool hasScenario) =>
            hasScenario
                ? DefaultIndentSize
                : 0;

        private static Some<string> WithIndentation(this in Some<string> prefixedStep, Step step, bool hasScenario)
        {
            var indentSize = hasScenario.ToIndentSizeStep() + step.ToIndentSizeStepAndBut();
            var indent = new string(' ', indentSize);
            return $"{indent}{prefixedStep.Value}";
        }

        private static Some<string> WithOutcomeDescribed(this in Some<string> prefixedStep, Outcome outcome) =>
            $"{prefixedStep.Value} [{outcome.ToOutcomeText()}]";

        private static Some<string> ToPrefix(this Step step) =>
            step.ToString();

        public static Some<string> ToDescription(this StepResult stepResult) =>
            new Option<string>().WithPrefix(stepResult.Step.ToPrefix())
                .WithIndentation(stepResult.Step, hasScenario: false)
                .WithOutcomeDescribed(stepResult.Outcome);

        private static Some<string> ToDescription(this StepOutcome stepOutcome, bool hasScenario) =>
            stepOutcome
                .Text
                .WithPrefix(stepOutcome.Step.ToPrefix())
                .WithIndentation(stepOutcome.Step, hasScenario)
                .WithOutcomeDescribed(stepOutcome.Outcome);

        public static IReadOnlyList<StepResult> ToResults(this IReadOnlyList<StepOutcome> outcomes, bool hasScenario) =>
            outcomes.Select(o => 
                    new StepResult(
                        o.Step,
                        o.Outcome,
                        o.Text.IfNone(null),
                        o.ToDescription(hasScenario)))
                .ToList();

        public static IReadOnlyList<StepOutcome> WithLatestStepOutcomeAs(this IReadOnlyList<StepOutcome> outcomes, Outcome outcome) =>
            outcomes.Select((stepOutcome, index) =>
            {
                var isLastItem = index == outcomes.Count - 1;
                return isLastItem
                    ? stepOutcome.ToStepOutcomeOfOutcome(outcome)
                    : stepOutcome;
            }).ToList();

        private static StepOutcome ToStepOutcomeOfOutcome(this StepOutcome stepOutcome, Outcome outcome) =>
            new StepOutcome(stepOutcome.Step, outcome, stepOutcome.Text);

        public static StepOutcome ToStepOutcome(this in Some<Title> title, Outcome outcome) =>
            new StepOutcome(title.Value.Step, outcome, title.Value.Text);

        public static Some<Title> ToTitle(this string title, Step step) =>
            new Title(step, title);

        private static bool ExceptionTypeNameIsInconclusive(this in Some<string> exceptionTypeName) =>
            string.Equals(exceptionTypeName, "InconclusiveException", StringComparison.InvariantCultureIgnoreCase) ||
            string.Equals(exceptionTypeName, "AssertInconclusiveException", StringComparison.InvariantCultureIgnoreCase) ||
            (exceptionTypeName.Value.StartsWith("Skippable") && exceptionTypeName.Value.EndsWith("Exception"));

        private static bool ExceptionIsInconclusive(this in Some<Exception> ex) =>
            ExceptionTypeNameIsInconclusive(ex.Value.GetType().Name);

        public static Outcome ToOutcome(this in Some<Exception> ex) =>
            ex.ExceptionIsInconclusive()
                ? Outcome.Inconclusive
                : Outcome.Fail;
    }
}

File: BddPipe/BddPipe.FSharp/StepResultExtensions.fs
namespace BddPipe.FSharp

open System
open BddPipe.FSharp.Model
open BddPipe.FSharp.StringExtensions

module internal StepResultExtensions =

    [<Literal>]
    let private DefaultIndentSize = 2

    let toOutcometext outcome =
        match outcome with
            | Pass -> "Passed"
            | Fail -> "Failed"
            | Inconclusive -> "Inconclusive"
            | NotRun -> "not run"

    let withIndendation step hasScenario prefixedStep =
        let toIndentSizeStepAndBut step =
            match step with
            | And | But -> DefaultIndentSize
            | _ -> 0

        let toIndentSizeStep hasScenario =
            if hasScenario then DefaultIndentSize else 0

        let indentSize = toIndentSizeStep hasScenario + toIndentSizeStepAndBut step
        let indent = new string(' ', indentSize)
        sprintf "%s%s" indent prefixedStep

    let withOutcomeDescribed outcome prefixedStep =
        sprintf "%s [%s]" prefixedStep (toOutcometext outcome)

    let private toPrefix (step: Step) =
        step.ToString()

    let toDescription (stepResult: StepResult) =
        None
            |> withPrefix (toPrefix stepResult.Step)
            |> withIndendation stepResult.Step false
            |> withOutcomeDescribed stepResult.Outcome

    let private toStepOutcomeDescription hasScenario (stepOutcome: StepOutcome) =
        stepOutcome.Text
            |> withPrefix (toPrefix stepOutcome.Step)
            |> withIndendation stepOutcome.Step hasScenario
            |> withOutcomeDescribed stepOutcome.Outcome

    let toResults (outcomes: StepOutcome list) hasScenario =
        outcomes
            |> Seq.map (fun o -> {
                Step = o.Step
                Outcome = o.Outcome
                Title = o.Text
                Description = o |> toStepOutcomeDescription hasScenario })
            |> Seq.toList

    let private toStepOutcomeOfOutcome (stepOutcome: StepOutcome) outcome = {
        Step = stepOutcome.Step
        Outcome = outcome
        Text = stepOutcome.Text
    }

    let withLatestStepOutcomeAs (outcomes: StepOutcome list) outcome =
        outcomes |> Seq.mapi (fun index stepOutcome ->
                    match index with
                        | i when i = outcomes.Length - 1 -> toStepOutcomeOfOutcome stepOutcome outcome
                        | _ -> stepOutcome
                    )
                 |> Seq.toList

    let toStepOutcome title outcome = {
        Step = title.Step
        Outcome = outcome
        Text = title.Text
    }

    let toTitle step text = {
        Step = step
        Text = text
    }

    let private exceptionTypeNameIsInconclusive exceptionTypeName =
        String.Equals(exceptionTypeName, "InconclusiveException", StringComparison.InvariantCultureIgnoreCase) ||
        String.Equals(exceptionTypeName, "AssertInconclusiveException", StringComparison.InvariantCultureIgnoreCase) ||
        (exceptionTypeName.StartsWith("Skippable") && exceptionTypeName.EndsWith("Exception"))

    let private exceptionIsInconclusive (ex: Exception) =
        exceptionTypeNameIsInconclusive (ex.GetType().Name)

    let toOutcome (ex: Exception) =
        match exceptionIsInconclusive ex with
            | true -> Inconclusive
            | false -> Fail
